### PR TITLE
Handle Composable Stable pools

### PR DESCRIPTION
**Pull request by alter-eggo**
# Description
We currently handle `ComposableStable` pools as if they were only an upgrade for the `PhantomStable` pool type and so the UI expects these pools to have a composition of only `LinearPools`. However, this is not the case, we have some `ComposableStable` pools which are just plain stable pools with regular tokens as pool tokens and so this pr handles this.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## How should this be tested?

Please provide instructions so we can test. Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

## Visual context

Please provide any relevant visual context for UI changes or additions. This could be static screenshots or a loom screencast.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have requested at least 2 reviews (If the PR is significant enough, use best judgement here)
- [x] I have commented my code where relevant, particularly in hard-to-understand areas
- [x] If package-lock.json has changes, it was intentional.
- [ ] The base of this PR is `master` if hotfix, `develop` if not

### CODE REVIEW

File: src/composables/pools/usePoolMigration.ts [hunk 1]
```diff
@@ -133,17 +133,22 @@ export function usePoolMigration(
...
+
+      case PoolMigrationType.STMATIC_POOL:
+      case PoolMigrationType.XMATIC_POOL:
+        return migrateStables;
+
+      case PoolMigrationType.MAI_POOL:
+        return migrateMaiUsd;
+
+      default:
+        return migrateBoostedPool;
```
Comment (garethfuller) [inline src/composables/pools/usePoolMigration.ts:151]: Again, not sure why pool migrations stuff is in this branch.

File: src/composables/queries/useUserPoolsQuery.ts [hunk 2]
```diff
@@ -77,9 +77,7 @@ export default function useUserPoolsQuery(
     });
 
     for (let i = 0; i < pools.length; i++) {
-      const isComposableStableLikePool = isComposableStableLike(
-        pools[i].poolType
-      );
+      const isComposableStableLikePool = isDeep(pools[i]);
```
Comment (garethfuller) [inline src/composables/queries/useUserPoolsQuery.ts:80]: I don't think const is needed anymore, was added to simplify code but function call is clear enough. Just change if statement blow to:
```ts
if (isDeep(pools[i]) {
```

File: src/composables/queries/useUserPoolsQuery.ts [hunk 3]
```diff
@@ -107,16 +105,15 @@ export default function useUserPoolsQuery(
 
     // TODO - cleanup and extract elsewhere in refactor
     for (let i = 0; i < decoratedPools.length; i++) {
-      const isComposableStableLikePool = isComposableStableLike(
-        decoratedPools[i].poolType
-      );
+      const isComposableStableLikePool = isDeep(decoratedPools[i]);
```
Comment (garethfuller) [inline src/composables/queries/useUserPoolsQuery.ts:108]: Same here, const no longer required.

File: src/composables/usePool.ts [hunk 4]
```diff
@@ -95,11 +113,21 @@ export function noInitLiquidity(pool: AnyPool): boolean {
   return bnum(pool?.onchain?.totalSupply || '0').eq(0);
 }
 
+export function checkPoolItselfTokenIndex(pool: Pool): number | void {
```
Comment (garethfuller) [inline src/composables/usePool.ts:116]: Change name to:
```suggestion
export function preMintedBptIndex(pool: Pool): number | void {
```

File: src/composables/usePool.ts [hunk 5]
```diff
@@ -95,11 +113,21 @@ export function noInitLiquidity(pool: AnyPool): boolean {
...
+  const index = pool.tokens.findIndex(token => token.address === pool.address);
+  if (index === -1) {
+    console.warn(`Pool itself token is not found in pool tokens list`);
+    return;
+  }
+
+  return index;
```
Comment (garethfuller) [inline src/composables/usePool.ts:123]: Cleanup
```suggestion
  return pool.tokens.findIndex(token => token.address === pool.address);
```

File: src/composables/usePool.ts [hunk 6]
```diff
@@ -287,11 +310,14 @@ export function usePool(pool: Ref<AnyPool> | Ref<undefined>) {
...
   );
+  const isDeepPool = computed(
+    (): boolean => !!pool.value && isDeep(pool.value)
+  );
   const isStableLikePool = computed(
     (): boolean => !!pool.value && isStableLike(pool.value.poolType)
   );
   const isComposableStableLikePool = computed(
-    (): boolean => !!pool.value && isComposableStableLike(pool.value.poolType)
+    (): boolean => !!pool.value && isDeep(pool.value)
```
Comment (garethfuller) [inline src/composables/usePool.ts:327]: This is incorrect, it should remain as it was:
```suggestion
    (): boolean => !!pool.value && isComposableStableLike(pool.value.poolType)
```

File: src/services/pool/calculator/calculator.sevice.ts [hunk 7]
```diff
@@ -249,7 +245,7 @@ export default class CalculatorService {
   }
 
   public get isComposableStableLikePool(): boolean {
-    return isComposableStableLike(this.pool.value.poolType);
+    return isDeep(this.pool.value);
```
Comment (garethfuller) [inline src/services/pool/calculator/calculator.sevice.ts:248]: This shouldn't have been changed, please revert:
```suggestion
    return isComposableStableLike(this.pool.value.poolType);
```

**Review by garethfuller — CHANGES_REQUESTED** [review]

File: src/components/contextual/pages/vebal/LMVoting/GaugesTable.vue [hunk 8]
```diff
@@ -132,8 +133,7 @@ const dataKey = computed(() => JSON.stringify(props.data));
  */
 function orderedTokenURIs(gauge: VotingGaugeWithVotes): string[] {
   const sortedTokens = orderedPoolTokens(
-    gauge.pool.poolType,
-    gauge.pool.address,
+    gauge.pool as unknown as Pool,
```
[the remark this answers is not in the document]
Comment (alter-eggo) [reply src/components/contextual/pages/vebal/LMVoting/GaugesTable.vue:136]: hm, it is possible now

File: src/components/forms/pool_actions/MigrateForm/constants.ts [hunk 9]
```diff
@@ -61,32 +61,41 @@ const MAINNET_POOL_MIGRATIONS: PoolMigrationInfo[] = [
...
+  {
+    type: PoolMigrationType.MAI_POOL,
+    fromPoolId: POOLS.IdsMap.mai4?.mai4 as string, // mai4
+    toPoolId: POOLS.IdsMap.mai4?.maiBbaUsd as string, // bbaMai
+    riskI18nLabels: [
+      `${riskI18nLabelPrefix}.loseUSDPeg`,
+      `${riskI18nLabelPrefix}.aaveStableExploit`,
+    ],
+  },
+];
```
[the remark this answers is not in the document]
Comment (alter-eggo) [reply src/components/forms/pool_actions/MigrateForm/constants.ts:92]: it was impossible to test migrations to polygon composable stable pools without changes from this pr, that's why I decided to make this changes here

File: src/composables/pools/usePoolMigration.ts (hunk 1 above)
Comment (alter-eggo) [reply src/composables/pools/usePoolMigration.ts:151]: answered in previous comment

**Comment by alter-eggo** [comment]
fixed all

File: src/composables/queries/useUserPoolsQuery.ts [hunk 10]
```diff
@@ -77,11 +77,7 @@ export default function useUserPoolsQuery(
     });
 
     for (let i = 0; i < pools.length; i++) {
-      const isComposableStableLikePool = isComposableStableLike(
-        pools[i].poolType
-      );
-
-      if (isComposableStableLikePool) {
+      if (isDeep(pools[i])) {
```
Comment (garethfuller) [inline src/composables/queries/useUserPoolsQuery.ts:80]: You were right actually, this needs to be a check if it's ComposableStable like.
```suggestion
      if (isComposableStableLike(pools[i].poolType)) {
```

File: src/services/pool/exchange/serializers/ExitParams.ts [hunk 11]
```diff
@@ -78,10 +100,22 @@ export default class ExitParams {
 
   private parseTokensOut(tokensOut: string[]): string[] {
     const nativeAsset = this.config.network.nativeAsset;
-
-    return tokensOut.map(address =>
+    const poolTokenItselfIndex = preMintedBptIndex(this.pool.value);
```
Comment (garethfuller) [inline src/services/pool/exchange/serializers/ExitParams.ts:103]: May as well rename this const too:
```suggestion
    const preMintedBptIndex = preMintedBptIndex(this.pool.value);
```

File: src/services/pool/exchange/serializers/JoinParams.ts [hunk 12]
```diff
@@ -44,16 +52,32 @@ export default class JoinParams {
       bptOut,
       this.pool.value?.onchain?.decimals || 18
     );
+    console.log('parsedBptOut', parsedBptOut.toString());
```
Comment (garethfuller) [inline src/services/pool/exchange/serializers/JoinParams.ts:55]: Remove this log.

**Review by garethfuller — CHANGES_REQUESTED** [review]
Just some additional minor things.

File: src/composables/queries/useUserPoolsQuery.ts (hunk 10 above)
Comment (alter-eggo) [reply src/composables/queries/useUserPoolsQuery.ts]: fixed here and below similar condition

**Review by garethfuller — APPROVED** [review]
Great, let's merge into pool migrations now.

**Pull request merged**